src/osvr/USBSerial/USBSerialDevInfo_MacOSX.h

```diff
@@ -51,51 +54,89 @@ namespace usbserial {
+            }
+            // Query only for serial ports
+            CFDictionarySetValue(classesToMatch, CFSTR(kIOSerialBSDTypeKey),
+                                 CFSTR(kIOSerialBSDAllTypes));
+            // Run the query
+            kernResult = IOServiceGetMatchingServices(
+                kIOMasterPortDefault, classesToMatch, matchingServices);
+            if (KERN_SUCCESS != kernResult) {
+                return -1;
+            }
+            return 0;
         }
-
-        /**
-         * Predicate that determines if the given vendor and product IDs match
-         * that of the provided USB serial device. If the @param vendorID or
-         * @param productID is @c boost::none, that parameter will match and ID.
-         *
-         * @param device the USB serial device to compare against
-         * @param vendorID optional vendor ID
-         * @param productID optional product ID
-         *
-         * @return true if the vendor and product IDs match, false otherwise
-         */
-        bool matches_ids(const USBSerialDevice &device,
-                         const boost::optional<uint16_t> &vendorID,
-                         const boost::optional<uint16_t> &productID) {
-            const bool vendor_matches =
-                (!vendorID || *vendorID == device.getVID());
-            const bool product_matches =
-                (!productID || *productID == device.getPID());
-
-            return (vendor_matches && product_matches);
-        }
-
-    } // end namespace
+    }
 
     std::vector<USBSerialDevice>
     getSerialDeviceList(boost::optional<uint16_t> vendorID,
                         boost::optional<uint16_t> productID) {
 
-        //FIXME: STUB
         std::vector<USBSerialDevice> devices;
+
+        io_iterator_t serialPortIterator;
+        io_object_t serialPortService;
+        // find all serial ports
+        if (findSerialPorts(&serialPortIterator) != 0) {
+            goto bailout;
+        }
+
+        // iterate over serial ports, getting vid and pid
+        while ((serialPortService = IOIteratorNext(serialPortIterator)) != 0) {
+            const CFNumberRef vidObj =
+                static_cast<CFNumberRef>(IORegistryEntrySearchCFProperty(
+                    serialPortService, kIOServicePlane, CFSTR("idVendor"), NULL,
+                    kIORegistryIterateRecursively | kIORegistryIterateParents));
+            const CFNumberRef pidObj =
+                static_cast<CFNumberRef>(IORegistryEntrySearchCFProperty(
+                    serialPortService, kIOServicePlane, CFSTR("idProduct"),
+                    NULL,
+                    kIORegistryIterateRecursively | kIORegistryIterateParents));
+            const CFStringRef bsdPathObj =
+                static_cast<CFStringRef>(IORegistryEntryCreateCFProperty(
+                    serialPortService, CFSTR(kIOCalloutDeviceKey),
+                    kCFAllocatorDefault, 0));
+
+            if (vidObj && pidObj && bsdPathObj) {
```

Comment:
I would probably invert this conditional and loop earlier. This prevents deep nesting. `if (!vidObj || !pidObj || !bsdPatchObj) continue;`